Add not working idl examples

ivanpauno opened the pull request:
I copied some idl examples from https://github.com/ros2/rosidl/blob/4b31e72cd7ce5b7e5b7f110472b063365e677132/rosidl_parser/test/, and added them to the lists here.

Interfaces are generated when [building test_msgs](https://github.com/ros2/rcl_interfaces/blob/a7ea280423e64f38d910e318d73a55bc22e71d42/test_msgs/CMakeLists.txt#L23-L32):
```
colcon build --symlink-install --packages-up-to test_msgs
```

I have build errors when I left any of the `.idl` files uncommented.

ivanpauno committed:
Add not working idl examples

Signed-off-by: ivanpauno <ivanpauno@ekumenlabs.com>

ivanpauno committed:
Replace rosidl_parser module name with test_msgs

Signed-off-by: ivanpauno <ivanpauno@ekumenlabs.com>

ivanpauno commented:
Error when commenting out all ".idl" files except of `MyMessage.idl`:

```
TypeError in template '/home/ivanpauno/ros2_ws/install/rosidl_generator_dds_idl/share/rosidl_generator_dds_idl/resource/msg.idl.em': %d format: a number is required, not str
TypeError when expanding 'idl.idl.em' into '/home/ivanpauno/ros2_ws/build/test_msgs/rosidl_generator_dds_idl/test_msgs/msg/dds_opensplice/MyMessage_.idl': %d format: a number is required, not str
Error processing idl file: /home/ivanpauno/ros2_ws/install/test_interface_files/share/test_interface_files/msg/MyMessage.idl
Traceback (most recent call last):
  File "/home/ivanpauno/ros2_ws/install/rosidl_generator_dds_idl/lib/rosidl_generator_dds_idl/rosidl_generator_dds_idl", line 41, in <module>
    sys.exit(main())
  File "/home/ivanpauno/ros2_ws/install/rosidl_generator_dds_idl/lib/rosidl_generator_dds_idl/rosidl_generator_dds_idl", line 36, in main
    args.additional_service_templates,
  File "/home/ivanpauno/ros2_ws/install/rosidl_generator_dds_idl/lib/python3.6/site-packages/rosidl_generator_dds_idl/__init__.py", line 51, in generate_dds_idl
    generate_files(generator_arguments_file, mapping, additional_context, keep_case=True)
  File "/home/ivanpauno/ros2_ws/install/rosidl_cmake/lib/python3.6/site-packages/rosidl_cmake/__init__.py", line 94, in generate_files
    raise(e)
  File "/home/ivanpauno/ros2_ws/install/rosidl_cmake/lib/python3.6/site-packages/rosidl_cmake/__init__.py", line 89, in generate_files
    post_process_callback=post_process_callback)
  File "/home/ivanpauno/ros2_ws/install/rosidl_cmake/lib/python3.6/site-packages/rosidl_cmake/__init__.py", line 148, in expand_template
    interpreter.string(template_content, template_path, locals=data)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2391, in string
    self.safe(scanner, True, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2401, in safe
    self.parse(scanner, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2421, in parse
    token.run(self, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 1425, in run
    interpreter.execute(self.code, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2596, in execute
    _exec(statements, self.globals, locals)
  File "<string>", line 7, in <module>
  File "/home/ivanpauno/ros2_ws/install/rosidl_cmake/lib/python3.6/site-packages/rosidl_cmake/__init__.py", line 197, in _expand_template
    interpreter.string(content, str(template_path), kwargs)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2391, in string
    self.safe(scanner, True, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2401, in safe
    self.parse(scanner, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2421, in parse
    token.run(self, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 1547, in run
    self.subrun(info[0][1], interpreter, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 1631, in subrun
    token.run(interpreter, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 1527, in run
    self.subrun(subtokens, interpreter, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 1631, in subrun
    token.run(interpreter, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 1333, in run
    result = interpreter.evaluate(self.testCode, locals)
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2573, in evaluate
    result = eval(expression, self.globals, locals)
  File "<string>", line 1, in <module>
  File "/home/ivanpauno/ros2_ws/install/rosidl_typesupport_opensplice_cpp/lib/python3.6/site-packages/rosidl_typesupport_opensplice_cpp/rosidl_generator_dds_idl_extension.py", line 24, in idl_typename
    typename = default_idl_typename(type_)
  File "/home/ivanpauno/ros2_ws/install/rosidl_generator_dds_idl/lib/python3.6/site-packages/rosidl_generator_dds_idl/__init__.py", line 86, in idl_typename
    typename += '<%d>' % (type_.maximum_size)
TypeError: %d format: a number is required, not str
Exception ignored in: <bound method Interpreter.__del__ of <empy pseudomodule/interpreter at 0x7ff1d42b4908>>
Traceback (most recent call last):
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2094, in __del__
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2157, in shutdown
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2647, in finalize
  File "/home/ivanpauno/.local/lib/python3.6/site-packages/em.py", line 2200, in push
AttributeError: '_io.TextIOWrapper' object has no attribute 'push'
make[2]: *** [rosidl_generator_dds_idl/test_msgs/msg/dds_opensplice/MyMessage_.idl] Error 1
make[1]: *** [CMakeFiles/test_msgs__dds_opensplice_idl.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
In file included from /home/ivanpauno/ros2_ws/build/test_msgs/rosidl_generator_c/test_msgs/msg/my_message__functions.h:19:0,
                 from /home/ivanpauno/ros2_ws/build/test_msgs/rosidl_generator_c/test_msgs/msg/my_message__functions.c:4:
/home/ivanpauno/ros2_ws/build/test_msgs/rosidl_generator_c/test_msgs/msg/my_message__struct.h:71:84: error: ‘UNSIGNED_LONG_CONSTANT’ undeclared here (not in a function)
   rosidl_parser__msg__MyMessage__constant_bounded_wstring_value__MAX_STRING_SIZE = UNSIGNED_LONG_CONSTANT
```

If I replace `wstring<UNSIGNED_LONG_CONSTANT> constant_bounded_wstring_value;` with `wstring<42> constant_bounded_wstring_value;` I get:
```
Starting >>> test_msgs
--- stderr: test_msgs                               
Traceback (most recent call last):
  File "/home/ivanpauno/ros2_ws/install/rosidl_typesupport_opensplice_cpp/lib/rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp", line 55, in <module>
    sys.exit(main())
  File "/home/ivanpauno/ros2_ws/install/rosidl_typesupport_opensplice_cpp/lib/rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp", line 50, in main
    args.idl_pp,
  File "/home/ivanpauno/ros2_ws/install/rosidl_typesupport_opensplice_cpp/lib/python3.6/site-packages/rosidl_typesupport_opensplice_cpp/__init__.py", line 74, in generate_dds_opensplice_cpp
    subprocess.check_call(cmd, cwd=folder)
  File "/usr/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/idlpp', '-I', '/home/ivanpauno/ros2_ws/build/test_msgs/rosidl_generator_dds_idl', '-I', '/home/ivanpauno/ros2_ws/install/builtin_interfaces/share', '-I', '/home/ivanpauno/ros2_ws/install/action_msgs/share', '-I', '/home/ivanpauno/ros2_ws/install/unique_identifier_msgs/share', '-I', '/usr/etc/opensplice/idlpp', '-S', '-l', 'cpp', '-N', '-d', '/home/ivanpauno/ros2_ws/build/test_msgs/rosidl_typesupport_opensplice_cpp/test_msgs/msg/dds_opensplice', '-o', 'maintain-include-namespace', 'MyMessage_.idl']' returned non-zero exit status 255.
make[2]: *** [rosidl_typesupport_opensplice_cpp/test_msgs/msg/dds_opensplice/my_message__type_support.cpp] Error 1
make[2]: *** Deleting file 'rosidl_typesupport_opensplice_cpp/test_msgs/msg/dds_opensplice/my_message__type_support.cpp'
make[1]: *** [CMakeFiles/test_msgs__rosidl_typesupport_opensplice_cpp.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
---
```

dirk-thomas commented:
I guess in the case of an upper string bound the constant name is not resolved into the actual constant value. If this only fails for the `rosidl_generator_dds_idl` generator it might need some change to handle this case (maybe just using `%s` is sufficient?).

The second error doesn't contain an actual error message. You might want to try running the OpenSplice code generator manually to see if it give any valuable error message (which we don't capture).

ivanpauno commented:
> I guess in the case of an upper string bound the constant name is not resolved into the actual constant value. If this only fails for the `rosidl_generator_dds_idl` generator it might need some change to handle this case (maybe just using `%s` is sufficient?).
> 
> The second error doesn't contain an actual error message. You might want to try running the OpenSplice code generator manually to see if it give any valuable error message (which we don't capture).

I hacked the subprocess call in `generate_dds_opensplice_cpp` to see the error message, here a detailed list of the errors I found (I started by replacing `UNSIGNED_LONG_CONSTANT` with 42 as commented [before](https://github.com/ros2/test_interface_files/pull/10#issuecomment-524956571)).
- `b"***DDS parse error L undefined at line: 21 (searching from scope 'MyMessage_Constants')\n"`
- After deleting module `MyMessage_Constants`, I get: `b"***DDS parse error key undefined at line: 20 (searching from scope 'MyMessage_')\n"`
- After deleting `@key`, I get (what to delete is not intuitive from the error, I just did trial and error until the message changed): `b"*** DDS error in file MyMessage_.idl: Type 'wchar' (defined in MyMessage_) unsupported near the token #pragma  keylist MyMessage_\n (line: 58, column: 0)\n"`
  It seems from http://download.prismtech.com/docs/Vortex/html/ospl/DDSTutorial/topics-etc.html that `OpenSplice` way of specifying keys is different.
- After deleting `wchar wchar_value;` I get `b"*** DDS error in file MyMessage_.idl: Type 'long double' (defined in MyMessage_) unsupported near the token #pragma  keylist MyMessage_\n (line: 57, column: 0)\n"`.
- After deleting `long double long_double_value;` ... it builds fine!!

dirk-thomas commented:
What `idl` file are you passing to the OpenSplice code generator? The one you wrote or the custom derived from that which uses OpenSplice compatible syntax (which should be located in `build/<pkgname>/rosidl_generator_dds_idl/<pkgname>/msg/dds_opensplice/<idlname>_.idl`)?

ivanpauno commented:
> What `idl` file are you passing to the OpenSplice code generator? The one you wrote or the custom derived from that which uses OpenSplice compatible syntax (which should be located in `build/<pkgname>/rosidl_generator_dds_idl/<pkgname>/msg/dds_opensplice/<idlname>_.idl`)?

I'm not calling `OpenSplice` code generator directly, I'm calling ros2 generator (https://github.com/ros2/rcl_interfaces/blob/a7ea280423e64f38d910e318d73a55bc22e71d42/test_msgs/CMakeLists.txt#L23-L32). I see the error message from `OpenSplice` generator because I modified [this call](https://github.com/ros2/rosidl_typesupport_opensplice/blob/ccfd3a89f04ae813aa0107f640f29f0c4e002bea/rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py#L74) to show the error message.
I didn't check the custom derived message generated from `rosidl_generator_dds_idl`, I just modified the [original .idl](https://github.com/ros2/test_interface_files/blob/19e5ead1b3fb71395cf1c66ab1c45de1aa9c6595/msg/MyMessage.idl) file that I'm adding in this PR .

ivanpauno commented:
> It seems from http://download.prismtech.com/docs/Vortex/html/ospl/DDSTutorial/topics-etc.html that OpenSplice way of specifying keys is different.

About this, the error only happens when using both `@range` and `@key` in the same member, but it works fine if you only use one (and actually, the generated opensplice specific`.idl` file looks wrong when you use both).

dirk-thomas commented:
OpenSplice doesn't support several parts of the current IDL spec:
* `long double`
* `wchar`
* `wstring`
* annotations starting with `@`

I am not sure if we should spend time on those cases with CycloneDDS likely replacing OpenSplice at some point in the (near?) future.

jacobperron commented:
@ivanpauno Friendly ping. Do you plan to continue this PR or should we close it?

It would be nice to have some examples of generating code from pure IDL files, but I don't know if they need to belong in `test_interfaces_files`. Unless we plan to use them in multiple places, I think we could also just add such files directly to `test_msgs`.

ivanpauno commented:
> Do you plan to continue this PR or should we close it?

Not really, feel free to close.

> It would be nice to have some examples of generating code from pure IDL files, but I don't know if they need to belong in test_interfaces_files. Unless we plan to use them in multiple places, I think we could also just add such files directly to test_msgs.

That would be nice.
Most of the errors above only happened with opensplice, so I'm not sure if they are still valid or not.

jacobperron commented:
I just tried building and hit the same error related to the wrong constant name:

```
build/test_msgs/rosidl_generator_c/test_msgs/msg/detail/my_message__struct.h:71:80: error: ‘UNSIGNED_LONG_CONSTANT’ undeclared here (not in a function)
   71 | g__MyMessage__constant_bounded_wstring_value__MAX_STRING_SIZE = UNSIGNED_LONG_CONSTANT
      |                                                                 ^~~~~~~~~~~~~~~~~~~~~~

make[2]: *** [CMakeFiles/test_msgs__rosidl_generator_c.dir/build.make:419: CMakeFiles/test_msgs__rosidl_generator_c.dir/rosidl_generator_c/test_msgs/msg/detail/my_message__functions.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:393: CMakeFiles/test_msgs__rosidl_generator_c.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

The code generator should either be substituting the numeric value of `UNSIGNED_LONG_CONSTANT` or used the fully-qualified name `test_msgs__msg__MyMessage__UNSIGNED_LONG_CONSTANT`. I'm not sure what is preferable.

Is rosidl the appropriate repo to open a bug report?

ivanpauno commented:
> Is rosidl the appropriate repo to open a bug report?

IIUC, https://github.com/ros2/rosidl_dds/tree/master/rosidl_generator_dds_idl is the correct place.

jacobperron commented:
Ticketed the issue: https://github.com/ros2/rosidl_dds/issues/53
